Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XLGPlusPlus/XLGToDBAnalyzer.cs

```diff
@@ -389,24 +470,155 @@ public override void DominoInvocation(DominoInvocationEventData data)
+                    else if (pipType == PipType.WriteFile)
+                    {
+                        var writeFilePip = (Pips.Operations.WriteFile)hydratedPip;
+                        xldbSpecificPip = writeFilePip.ToWriteFile(PathTable, xldbPip);
+                    }
+                    else if (pipType == PipType.Process)
+                    {
+                        var processPip = (Pips.Operations.Process)hydratedPip;
+                        xldbSpecificPip = processPip.ToProcessPip(PathTable, xldbPip);
+                    }
+
+                    var pipSemistableQuery = new PipQuerySemiStableHash()
+                    {
+                        SemiStableHash = hydratedPip.SemiStableHash,
+                    };
+
+                    var pipSemistableValue = new PipValueSemiStableHash()
+                    {
+                        PipId = hydratedPip.PipId.Value,
+                    };
+
+                    var pipIdQuery = new PipQueryPipId()
+                    {
+                        PipId = hydratedPip.PipId.Value,
+                        PipType = (Xldb.PipType)pipType
+                    };
+
+                    // If the SemiStableHash != 0, then we want to create the SemistableHash -> PipId indirection.
+                    // Else we do not want that since the key would no longer be unique
+                    if (hydratedPip.SemiStableHash != 0)
+                    {
+                        pipSemistableKeys.Add(pipSemistableQuery.ToByteArray());
+                        pipSemistableValues.Add(pipSemistableValue.ToByteArray());
+                    }
+
+                    pipPipIdKeys.Add(pipIdQuery.ToByteArray());
+                    pipValues.Add(xldbSpecificPip.ToByteArray());
+                }
+
+                // Write the rest of the batched pips to the db
+                WriteBatchToDb(pipSemistableKeys, pipSemistableValues, XldbDataStore.PipColumnFamilyName);
+                WriteBatchToDb(pipPipIdKeys, pipValues, XldbDataStore.PipColumnFamilyName);
+            });
         }
 
         /// <summary>
         /// Write a key/value pair to the db
         /// </summary>
-        public void WriteToDb(byte[] key, byte[] value)
+        public void WriteToDb(byte[] key, byte[] value, string columnFamilyName = null)
+        {
+            Analysis.IgnoreResult(
+                m_accessor.Use(database =>
+                {
+                    database.Put(key, value, columnFamilyName);
+                })
+            );
+        }
+
+        /// <summary>
+        /// Write batches of key/value pairs to the db
+        /// </summary>
+        public void WriteBatchToDb(IEnumerable<byte[]> keys, IEnumerable<byte[]> values, string columnFamilyName = null)
         {
             Analysis.IgnoreResult(
                 m_accessor.Use(database =>
                 {
-                    database.Put(key, value);
+                    database.ApplyBatch(keys, values, columnFamilyName);

```

Comment:
IKeyValueStore no longer has the [contravariance](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/in-generic-modifier) requirement on `TKey`, so `ApplyBatch` can be changed to take a `Dictionary<byte[], byte[]>`, which would at the same time let you to avoid using so many lists in the method that calls this.

Reply:
As discussed offline this can be done in a different PR since several places/functions would need to be changed that use ApplyBatch

---
In reply to: [314126225](https://github.com/microsoft/BuildXL/pull/724#discussion_r314126225) [](ancestors = 314126225)